composer/algorithms/gyro_dropout/gyro_dropout.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2022  Gihyun Park, Junyeol Lee, and Jiwon Seo
+# SPDX-License-Identifier: Apache-2.0
+import logging
+import warnings
+from typing import Dict, Optional, Sequence, Type, Union
+
+import torch
+import numpy as np
+from torch.optim import Optimizer
+
+from composer.algorithms.warnings import NoEffectWarning
+from composer.core import Algorithm, Event, State
+from composer.loggers import Logger
+from composer.utils import module_surgery
+
+log = logging.getLogger(__name__)
+
+
+class GyroDropoutLayer(torch.nn.Module):
+    def __init__(self, iters_per_epoch: int, max_epoch: int, p:float, sigma:int, tau:int):
+        super(GyroDropoutLayer, self).__init__()    
+
+        self.sigma = sigma
+        self.tau = tau
+        self.p = p
+        self.preselect_masks = None
+        self.dropout_mask = None
+        self.training_step = 0
+        self.iter_num = 0
+        self.max_epoch = max_epoch
+        self.iters_per_epoch = iters_per_epoch
+        
+    def forward(self, x):
+        if self.training:
+            if self.training_step == 0:
+                is_cuda_tensor = x.is_cuda
+
+                if is_cuda_tensor:
+                    self.preselect_masks = (torch.rand(self.sigma, x.shape[1]) > self.p).float().to("cuda")
+                else:
+                    self.preselect_masks = (torch.rand(self.sigma, x.shape[1]) > self.p).float()
+                
+                # Below simplified from: (iters_per_epoch*max_epoch*batch_size/sigma) / (batch_size/self.tau) 
+                self.iter_num = int(self.iters_per_epoch * self.max_epoch / self.sigma) * self.tau
+                
+            if self.training_step % self.iter_num == 0:
+                pick_idx = np.random.choice(self.sigma, self.tau)
+                self.selected_masks = self.preselect_masks[pick_idx]
+
+            self.dropout_mask = torch.repeat_interleave(self.selected_masks, x.shape[0] // self.tau, dim=0)
+
+            self.training_step += 1          
+
+            return x * self.dropout_mask * (1 / (1 - self.p))
+        else:
+            return x
+
+
+def from_Dropout(p: float, sigma: int, tau: int, num_iterations: int, epoch:int, layer: torch.nn.Module, module_index: int):
+    """Defines a replacement policy from a `torch.nn.Dropout` to a 'GyroDropout`"""
+
+    return GyroDropoutLayer(p, sigma, tau, num_iterations, epoch)
+
+
+def apply_gyro_dropout(model: torch.nn.Module, 
+                       iters_per_epoch: int,
```

Comment:
Algorithms being dependent on iters_per_epoch and max_epoch is not ideal as we will have to change these parameters for this algorithm manually whenever training duration changes. You can access training progress from state. cc: @mvpatel2000